app/assets/javascripts/govuk_publishing_components/components/autocomplete.js

```diff
@@ -0,0 +1,561 @@
+//= require ../vendor/polyfills/closest.js
+/* global fetch */
+
+window.GOVUK = window.GOVUK || {}
+window.GOVUK.Modules = window.GOVUK.Modules || {}
+window.GOVUK.Modules.GovukAutocomplete = window.GOVUKFrontend.Autocomplete;
+
+(function (Modules) {
+  'use strict'
+
+  var util = {
+    getRelativePosition: function (element1, element2) {
+      var position1 = element1.getBoundingClientRect()
+      var position2 = element2.getBoundingClientRect()
+
+      var positionAbove =
+        /* 1 */ position1.bottom + position2.height > window.innerHeight &&
+        /* 2 */ window.innerHeight - position1.bottom < position1.top &&
+        /* 3 */ window.pageYOffset + position1.top - position2.height > 0
+
+      return positionAbove ? 'above' : 'below'
+    },
+    debounce: function (func, wait, immediate) {
+      let timeout
+
+      return function executedFunction () {
+        var context = this
+        var args = arguments
+
+        var later = function () {
+          timeout = null
+          if (!immediate) func.apply(context, args)
+        }
+
+        var callNow = immediate && !timeout
+        clearTimeout(timeout)
+        timeout = setTimeout(later, wait)
+
+        if (callNow) func.apply(context, args)
+      }
+    },
+    getAriaLabel: function (labelStr) {
+      if (labelStr?.length) {
+        var isLabelId = labelStr.startsWith('#')
+
+        return {
+          attribute: isLabelId ? 'aria-labelledby' : 'aria-label',
+          content: isLabelId ? labelStr.substring(1) : labelStr
+        }
+      }
+    },
+    isPromise: function (value) {
+      return Boolean(value && typeof value.then === 'function')
+    },
+    isFunction: function (value) {
+      return Boolean(typeof value === 'function')
+    },
+    isArray: function (string) {
+      try {
+        return Array.isArray(JSON.parse(string))
+      } catch (e) {
+        return false
+      }
+    }
+  }
+
+  // Creates a props object with overridden toString function. toString returns an attributes
+  // string in the format: `key1='value1' key2='value2'` for easy use in an HTML string.
+  function Props (index, selectedIndex, baseClass) {
+    this.id = `${baseClass}-result-${index}`
+    this.class = `${baseClass}__result`
+    this['data-result-index'] = index
+    this.role = 'option'
+    if (index === selectedIndex) {
+      this['aria-selected'] = 'true'
+    }
+  }
+
+  Props.prototype.toString = function () {
+    return Object.keys(this).reduce(
+      (str, key) => `${str} ${key}='${this[key]}'`,
+      ''
+    )
+  }
+
+  function AutocompleteCore (
+    search,
+    autoSelect,
+    setValue,
+    setAttribute,
+    onUpdate,
+    onSubmit,
+    onShow,
+    onHide,
+    onLoading,
+    onLoaded,
+    submitOnSelect
+  ) {
+    this.value = ''
+    this.searchCounter = 0
+    this.results = []
+    this.selectedIndex = -1
+    this.selectedResult = null
+
+    switch (true) {
```

Comment:
Haven't seen a switch statement based on `true` before - would this be cleaner as an if statement? e.g.

```JS
if (util.isFunction(search)) {
  //
} else if (util.isArray(search)) {
  //
} else {
  //
}
```